assets/javascripts/needlediff.js

```diff
@@ -138,7 +133,40 @@ NeedleDiff.prototype.draw = function () {
         if (!this.fullNeedleImg) {
           // draw matching part of needle image
           this.ctx.strokeStyle = NeedleDiff.strokecolor(a.type);
-          this.ctx.drawImage(this.needleImg, orig.xpos, orig.ypos, usedWith, a.height, x, a.ypos, usedWith, a.height);
+          if (this.needleImg) {
+            this.ctx.drawImage(this.needleImg, orig.xpos, orig.ypos, usedWith, a.height, x, a.ypos, usedWith, a.height);
+          } else {
+            // Calculate font size with 1px margin
+            var ocrLinesFontSize = 14;
+            var ocrLines = a['ocr_str'].split('\n');
+            while ((ocrLines.length * ocrLinesFontSize + (ocrLines.length - 1)) > (a.height - 2)) {
+              ocrLinesFontSize--;
+            }
+            this.ctx.font = 'bold ' + ocrLinesFontSize + 'px Arial';
+            for (line of ocrLines) {
+              var ocrLinesWidth = this.ctx.measureText(line).width;
+              while (ocrLinesWidth > (a.width - 2)) {
+                ocrLinesFontSize--;
+                this.ctx.font = 'bold ' + ocrLinesFontSize + 'px Arial';
+                ocrLinesWidth = this.ctx.measureText(line).width;
+              }
+            }
+
+            // Write OCR lines to Canvas context line by line
+            var ocrTxtXPos = Math.ceil(a.xpos + a.width / 2);
```

Comment:
It would generally be nice to use `const`/`let` in new code, e.g.

```suggestion
            const ocrTxtXPos = Math.ceil(a.xpos + a.width / 2);
```

This counts also for all the other occurences of `var …` (in new code).